crates/store/src/db/models/queries/transactions.rs

```diff
@@ -126,14 +173,191 @@ pub struct TransactionSummaryRowInsert {
+        let chunk =
+            SelectDsl::select(schema::transactions::table, TransactionRecordRaw::as_select())
+                .filter(schema::transactions::block_num.gt(block_range.start().to_raw_sql()))
+                .filter(schema::transactions::block_num.le(block_range.end().to_raw_sql()))
+                .filter(schema::transactions::account_id.eq_any(&desired_account_ids))
+                .order((schema::transactions::block_num.asc(),))
+                .offset(offset)
+                .limit(CHUNK_SIZE)
+                .load::<TransactionRecordRaw>(conn)
+                .map_err(DatabaseError::from)?;
+
+        // If no more data, we're done
+        if chunk.is_empty() {
+            break;
+        }
+
+        // Calculate the size of this chunk
+        let chunk_size: i64 = chunk.iter().map(|tx| tx.size_in_bytes).sum();
+
+        // If adding this chunk would exceed the limit, we need to be more careful
+        if current_size + chunk_size > MAX_PAYLOAD_BYTES as i64 {
+            // Add transactions from this chunk one by one until we hit the limit
+            for tx in chunk {
+                if current_size + tx.size_in_bytes <= MAX_PAYLOAD_BYTES as i64 {
+                    current_size += tx.size_in_bytes;
+                    all_transactions.push(tx);
+                } else {
+                    // Can't fit this transaction, stop here
+                    break;
+                }
+            }
+            break;
+        }
+
+        // The entire chunk fits, add it all
+        current_size += chunk_size;
+        let chunk_len = chunk.len() as i64;
+        all_transactions.extend(chunk);
+        offset += CHUNK_SIZE;
+
+        // If we got fewer transactions than requested, we've reached the end
+        if chunk_len < CHUNK_SIZE {
+            break;
         }
     }
+
+    // Ensure block consistency: remove complete blocks from the end until we're under the size
+    // limit
+    let mut filtered_raw = all_transactions;
+    while !filtered_raw.is_empty() {
+        let total_size: i64 = filtered_raw.iter().map(|tx| tx.size_in_bytes).sum();
+
+        // If we're under the limit, we're done
+        if total_size <= MAX_PAYLOAD_BYTES as i64 {
+            break;
+        }
+
+        // Find the last block number and remove all transactions from that block
+        if let Some(last_tx) = filtered_raw.last() {
+            let last_block_num = last_tx.block_num;
+            filtered_raw.retain(|tx| tx.block_num != last_block_num);
+        } else {
+            // This shouldn't happen since we check !filtered_raw.is_empty() above, but be safe
+            break;
+        }
+    }
+
+    // Convert to transaction records
+    let headers: Vec<crate::db::TransactionRecord> = filtered_raw
+        .into_iter()
+        .map(std::convert::TryInto::try_into)
+        .collect::<Result<Vec<_>, _>>()?;
+
+    let last_included_block = headers.last().map_or(*block_range.start(), |tx| tx.block_num);
```

Comment:
Can we not use the same approach for filtering out the last block as we did for other sync endpoints (e.g., sync storage map and sync asset vault - where we pass over the result set only once)?

Reply:
The same as in requesting something like `MAX_ROWS + 1` and then checking if `result[last].block_num != result[last - 1].block_num`, I didn't take that path because MAX_ROWS is different to calculate precisely due to the notes variability.